File: server/src/services/email-verification.service.ts
import { randomInt } from 'node:crypto'
import { logger } from '../lib/logger.js'
import { queryDb } from '../lib/db.js'
import { isEmailConfigured, sendVerificationEmail } from '../lib/email.js'

const CODE_LENGTH = 6
const CODE_TTL_MS = 5 * 60 * 1_000
const MAX_PENDING_PER_EMAIL = 5

function generateCode(): string {
  let code = ''
  for (let i = 0; i < CODE_LENGTH; i += 1) {
    code += randomInt(0, 10).toString()
  }
  return code
}

function normalizeEmail(email: string): string {
  return email.trim().toLowerCase()
}

export async function sendVerificationCode(email: string): Promise<void> {
  const normalizedEmail = normalizeEmail(email)

  if (!isEmailConfigured()) {
    throw new Error('邮件服务未配置，无法发送验证码')
  }

  const pendingCount = await queryDb<{ count: string }>(
    `select count(*) as count
       from email_verification_codes
      where email = $1
        and used = false
        and expires_at > now()`,
    [normalizedEmail],
  )

  const count = Number(pendingCount.rows[0]?.count ?? 0)
  if (count >= MAX_PENDING_PER_EMAIL) {
    throw new Error('验证码发送过于频繁，请稍后再试')
  }

  const code = generateCode()
  const expiresAt = new Date(Date.now() + CODE_TTL_MS)

  await queryDb(
    `insert into email_verification_codes (email, code, expires_at)
     values ($1, $2, $3)`,
    [normalizedEmail, code, expiresAt],
  )

  logger.info({ email: normalizedEmail }, 'Verification code generated')

  await sendVerificationEmail(normalizedEmail, code)
}

export async function verifyCode(email: string, code: string): Promise<boolean> {
  const normalizedEmail = normalizeEmail(email)

  const result = await queryDb<{ id: string }>(
    `select id
       from email_verification_codes
      where email = $1
        and code = $2
        and used = false
        and expires_at > now()
      order by created_at desc
      limit 1`,
    [normalizedEmail, code],
  )

  const row = result.rows[0]
  if (!row) {
    return false
  }

  await queryDb(
    `update email_verification_codes
        set used = true
      where id = $1`,
    [row.id],
  )

  return true
}

export async function cleanupExpiredCodes(): Promise<number> {
  const result = await queryDb(
    `delete from email_verification_codes
      where expires_at < now()`,
  )

  return result.rowCount ?? 0
}

File: platform-java/services/identity-service/src/main/java/com/grassland/identity/security/EmailVerificationService.java
package com.grassland.identity.security;

import java.security.SecureRandom;
import org.springframework.r2dbc.core.DatabaseClient;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
public class EmailVerificationService {
    private static final int CODE_LENGTH = 6;
    private static final long CODE_TTL_MINUTES = 5;
    private static final int MAX_PENDING = 5;
    private final SecureRandom random = new SecureRandom();
    private final DatabaseClient db;

    public EmailVerificationService(DatabaseClient db) { this.db = db; }

    public Mono<String> createCode(String email) {
        String normalized = email == null ? null : email.trim().toLowerCase();
        return db.sql("SELECT count(*)::int AS n FROM email_verification_codes WHERE email = :email AND used = false AND expires_at > now()")
            .bind("email", normalized).map(r -> r.get("n", Integer.class)).one()
            .flatMap(count -> {
                if (count != null && count >= MAX_PENDING) {
                    return Mono.error(new IllegalStateException("验证码发送过于频繁，请稍后再试"));
                }
                String code = generateCode();
                return db.sql("INSERT INTO email_verification_codes(email, code, expires_at) VALUES (:email, :code, now() + interval '"
                    + CODE_TTL_MINUTES + " minutes')")
                    .bind("email", normalized).bind("code", code).then().thenReturn(code);
            });
    }

    public Mono<Boolean> verifyAndConsume(String email, String code) {
        String normalized = email == null ? null : email.trim().toLowerCase();
        return db.sql("SELECT id FROM email_verification_codes WHERE email = :email AND code = :code AND used = false AND expires_at > now() ORDER BY created_at DESC LIMIT 1")
            .bind("email", normalized).bind("code", code).map(r -> r.get("id", String.class)).one()
            .flatMap(id -> db.sql("UPDATE email_verification_codes SET used = true WHERE id = CAST(:id AS uuid)")
                .bind("id", id).then().thenReturn(true))
            .defaultIfEmpty(false);
    }

    private String generateCode() {
        StringBuilder sb = new StringBuilder(CODE_LENGTH);
        for (int i = 0; i < CODE_LENGTH; i++) sb.append(random.nextInt(10));
        return sb.toString();
    }
}
